mirgecom/logging_quantities.py

```diff
@@ -316,4 +319,40 @@ def __call__(self) -> float:
         """Return the memory usage in MByte."""
         return self.process.memory_info()[0] / 1024 / 1024
 
+
+class DeviceMemoryUsage(LogQuantity):
+    """Logging support for GPU memory usage (Nvidia only currently)."""
+
+    def __init__(self, name: str = None):
+
+        if name is None:
+            name = "memory_usage_gpu"
+
+        super().__init__(name, "MByte", description="Memory usage (GPU)")
+
+        import ctypes
+        self.free = ctypes.c_size_t()
+        self.total = ctypes.c_size_t()
+
+        try:
+            # Use libcudart, not libcuda, since the latter requires a CUDA
+            # device context which interferes with pyopencl.
```

Comment:
Argh. I forgot the implicit context state in CUDA.

cudart only appears to solve your problem, but in actuality it does not. It activates and uses the device's primary context, which you could also get using `cuDevicePrimaryCtxReset` from the driver API. Neither of the two is the correct approach IMO, because it will cause expensive thrashing of the CUDA context stack, as discussed [here](https://github.com/pocl/pocl/issues/908).

IMO, the better approach is to poke a hole into pocl to allow it to tell us its CUDA (driver API) context and activate that if it isn't.

Reply:
Which metric would that effect? I tested this and couldn't see much difference, neither in the abosulte time it takes to log, nor in the nvprof output.

Measuring GPU memory disabled:

```
     API calls:    23.88%  3.08811s    435902  7.0840us  6.7060us  5.1683ms  cuEventElapsedTime
                   20.73%  2.68161s    207539  12.920us  11.599us  5.0780ms  cuLaunchKernel
                   14.79%  1.91251s    435906  4.3870us  2.7140us  5.0748ms  cuEventRecord
                   12.31%  1.59209s        34  46.826ms  43.578ms  56.050ms  cuModuleLoad
                    7.49%  968.30ms    217955  4.4420us  4.1520us  1.3048ms  cuEventSynchronize
                    6.51%  841.39ms         4  210.35ms  186.13ms  249.30ms  cuCtxCreate
                    5.45%  705.45ms    435906  1.6180us  1.0960us  5.3724ms  cuEventCreate
                    3.29%  425.04ms    431608     984ns     750ns  679.57us  cuEventDestroy
                    2.40%  310.74ms    436233     712ns     468ns  1.2547ms  cuCtxSetCurrent
                    1.43%  184.60ms    207539     889ns     755ns  155.32us  cuModuleGetGlobal
                    0.63%  81.532ms      4813  16.939us  14.470us  32.421us  cuMemcpyDtoDAsync
                    0.40%  51.903ms      3200  16.219us  14.513us  27.238us  cuMemsetD8Async
                    0.37%  47.474ms      2020  23.501us  20.991us  150.88us  cuMemcpyDtoHAsync
                    0.30%  38.169ms       296  128.95us  12.470us  31.554ms  cuMemAlloc
                    0.01%  1.3909ms       400  3.4770us     945ns  24.977us  cuPointerGetAttributes
                    0.01%  1.0937ms       104  10.516us     233ns  257.41us  cuDeviceGetAttribute
                    0.01%  963.02us        83  11.602us  9.1810us  20.597us  cuMemcpyHtoDAsync
                    0.00%  238.76us         4  59.690us  54.447us  73.446us  cuDeviceGetName
                    0.00%  228.46us         4  57.116us  32.231us  91.090us  cuMemGetInfo
                    0.00%  43.787us        34  1.2870us  1.1620us  1.3980us  cuModuleGetFunction
                    0.00%  11.626us         1  11.626us  11.626us  11.626us  cuStreamCreate
                    0.00%  9.7690us         4  2.4420us  1.4690us  3.9040us  cuDeviceGetPCIBusId
                    0.00%  5.3440us         8     668ns     408ns  1.3710us  cuDeviceGet
                    0.00%  1.5030us         3     501ns     424ns     595ns  cuDeviceGetCount
```

Measuring GPU memory enabled:

```
      API calls:   23.38%  3.08273s    435818  7.0730us  6.7160us  1.4709ms  cuEventElapsedTime
                   20.96%  2.76403s    207539  13.318us  11.811us  1.4885ms  cuLaunchKernel
                   15.64%  2.06182s    435822  4.7300us  3.1850us  1.4586ms  cuEventRecord
                   12.14%  1.60038s        34  47.070ms  43.905ms  56.049ms  cuModuleLoad
                    7.34%  967.97ms    217913  4.4420us  4.1790us  1.2762ms  cuEventSynchronize
                    6.42%  846.38ms         4  211.60ms  185.64ms  254.87ms  cuCtxCreate
                    5.35%  705.56ms    435822  1.6180us  1.0460us  1.7627ms  cuEventCreate
                    3.29%  433.73ms    431524  1.0050us     752ns  5.2634ms  cuEventDestroy
                    2.42%  319.42ms    436107     732ns     476ns  2.1917ms  cuCtxSetCurrent
                    1.41%  186.38ms    207539     898ns     748ns  153.86us  cuModuleGetGlobal
                    0.64%  83.937ms      4813  17.439us  15.103us  148.67us  cuMemcpyDtoDAsync
                    0.40%  53.267ms      3200  16.645us  15.068us  29.835us  cuMemsetD8Async
                    0.37%  48.224ms      2020  23.873us  21.375us  45.160us  cuMemcpyDtoHAsync
                    0.11%  13.928ms       100  139.28us  78.883us  367.41us  cudaMemGetInfo
                    0.05%  6.2256ms       254  24.510us  12.706us  799.58us  cuMemAlloc
                    0.03%  3.7748ms       480  7.8640us     256ns  260.14us  cuDeviceGetAttribute
                    0.03%  3.6379ms         4  909.48us  897.64us  940.43us  cuDeviceTotalMem
                    0.01%  1.4131ms       400  3.5320us     920ns  23.460us  cuPointerGetAttributes
                    0.01%  986.71us        83  11.888us  9.4220us  21.753us  cuMemcpyHtoDAsync
                    0.00%  498.65us         8  62.331us  51.484us  95.170us  cuDeviceGetName
                    0.00%  239.97us         4  59.991us  31.535us  102.31us  cuMemGetInfo
                    0.00%  45.195us        34  1.3290us  1.2070us  1.4430us  cuModuleGetFunction
                    0.00%  10.083us         1  10.083us  10.083us  10.083us  cuStreamCreate
                    0.00%  9.3180us         4  2.3290us  1.3890us  3.8220us  cuDeviceGetPCIBusId
                    0.00%  8.2460us        12     687ns     480ns  1.5370us  cuDeviceGet
                    0.00%  2.5220us         4     630ns     561ns     752ns  cuDeviceGetUuid
                    0.00%  2.0630us         4     515ns     418ns     604ns  cuDeviceGetCount
                    0.00%  1.3010us         1  1.3010us  1.3010us  1.3010us  cuInit
                    0.00%     988ns         1     988ns     988ns     988ns  cuDriverGetVersion
```

Reply:
- In the 13.928ms it takes to call `cudaMemGetInfo`. Without a need to switch contexts, that might be cheaper.
- Also, the  319.42ms spent in calls to `cuCtxSetCurrent` that are here that I'm trying to convince the pocl folks are not necessary.
- IMO calling both the CUDA runtime and driver API in the same program is bad engineering practice and, in my experience, connected to reliability problems.

Reply:
Based on my experiment here, using `cuMemGetInfo` might be futile... https://gist.github.com/f0k/63a664160d016a491b2cbea15913d549#gistcomment-3645873

Reply:
Fixed. See https://gist.github.com/f0k/63a664160d016a491b2cbea15913d549#gistcomment-3654335 for details.